libnetwork: node-discovery: fix error-handling, and some refactor

thaJeztah opened the pull request:
### libnetwork: handleDriverTableEvent: reduce some duplication

I guess the attempt here was to not duplicate code, but in doing so,
there was actually more code-duplication than necessary.


### libnetwork: Controller.handleNodeTableEvent, handleEpTableEvent move vars

move variables closer to where they're used, and rename them where appropriate.

### libnetwork: Controller.handleNodeTableEvent: skip networkdb.UpdateEvent

Controller.handleNodeTableEvent was logging an error if it would handle
a `networkdb.UpdateEvent`, however, the function would still continue
processing the event, and attempt to unmarshal an empty `[]byte` to a
`networkdb.NodeAddr` (resulting in an `unexpected end of JSON input` error).

If it _would_ proceed (i.e., if the unmarshal would not error), it would;

- call `Controller.processNodeDiscovery` (with an empty list of IPs)
- which would iterate over all registered drivers, calling `Controller.pushNodeDiscovery`
  for drivers that implement `discoverapi.Discover`.
- ultimately ending up in `Controller.pushNodeDiscovery` returning early
  because no IPs were provided.

Based on the above, it looks like returning early here is the right thing
to do.

### libnetwork: Controller.pushNodeDiscovery: return early

Skip fetching the agent and parsing its IP-address if we don't use it.

### libnetwork: Controller.handleNodeTableEvent: log event-type in errors

Allow discovering what kind of event was received if an error occurs.

### libnetwork: Controller.pushNodeDiscovery: take single IP, ignore nil

The signature to take multiple addresses originates in [libnetwork/188][1]
("Libnetwork Host Discovery using Swarm Discovery pkg), which initially
had a `diff` function as part of the `hostdiscovery` package, which would
produce a list of changes in IP-addresses (added, removed).

Discovery with external k/v stores was removed, and the `hostdiscovery`
package was removed in 00f9b23c3ae1614b648e9da1c6423c381f58153e (https://github.com/moby/moby/pull/42247), and
current code always pass a single IP-address (if any).

The Controller.pushNodeDiscovery method contained a check whether the
passed list of node-addresses was nil, however, this was never the
case because `Controller.handleNodeTableEvent` would unconditionally
[initialize a slice][2] (even if the event had a `nil` IP-address).

In case of a `nil` IP-address, this address would be;

- compared with the agent's advertiseAddr to determine if this is `self`
- encoded as a string, and sent as `NodeDiscoveryData.Address`. This string
  would be a literal `"<nil>"`.

While I don't know if there's other implementations that treat this information
different, looking at the [overlay implementation of `driver.DiscoverNew`][3],
it would check the field to be an empty string, and otherwise produce an
error (which wouldn't apply if it contained `"<nil>"`), and set the
driver's advertiseAddress to this value, effectively un-setting it
(`net.ParseIP()` silently ignores invalid values), or producing an error
(`netip.ParseAddr()` will error).

For completeness: `driver.DiscoverDelete` has no implementation in the
overlay driver, but both `driver.DiscoverNew` and `driver.DiscoverDelete`
are both implemented by the `remote` driver, which sends them to the
remote "as-is", so it will depend on the receiving end to handle.

So based on the overlay implementation, my assumption is that an empty
node-address should be considered _invalid_ (both for `DiscoverDelete`
and `DiscoverNew`), so let's return early if we encounter one, but log
a warning to discover cases where this happened.

[1]: https://github.com/moby/libnetwork/commit/ac4d7b61367e34ac341c2171d70d8b6a0a0fbecf
[2]: https://github.com/moby/moby/blob/5b53ddfcdd1c0721f875d3e237078ab7de891a57/libnetwork/agent.go#L905
[3]: https://github.com/moby/moby/blob/5b53ddfcdd1c0721f875d3e237078ab7de891a57/libnetwork/drivers/overlay/overlay.go#L92-L97


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
libnetwork: handleDriverTableEvent: reduce some duplication

I guess the attempt here was to not duplicate code, but in doing so,
there was actually more code-duplication than necessary.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: Controller.handleNodeTableEvent, handleEpTableEvent move vars

move variables closer to where they're used, and rename them
where appropriate.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: Controller.handleNodeTableEvent: skip networkdb.UpdateEvent

Controller.handleNodeTableEvent was logging an error if it would handle
a `networkdb.UpdateEvent`, however, the function would still continue
processing the event, and attempt to unmarshal an empty `[]byte` to a
`networkdb.NodeAddr` (resulting in an `unexpected end of JSON input` error).

If it _would_ proceed (i.e., if the unmarshal would not error), it would;

- call `Controller.processNodeDiscovery` (with an empty list of IPs)
- which would iterate over all registered drivers, calling `Controller.pushNodeDiscovery`
  for drivers that implement `discoverapi.Discover`.
- ultimately ending up in `Controller.pushNodeDiscovery` returning early
  because no IPs were provided.

Based on the above, it looks like returning early here is the right thing
to do.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: Controller.pushNodeDiscovery: return early

Skip fetching the agent and parsing its IP-address if we don't use it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: Controller.handleNodeTableEvent: log event-type in errors

Allow discovering what kind of event was received if an error occurs.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: Controller.pushNodeDiscovery: take single IP, ignore nil

The signature to take multiple addresses originates in [libnetwork/188][1]
("Libnetwork Host Discovery using Swarm Discovery pkg), which initially
had a `diff` function as part of the `hostdiscovery` package, which would
produce a list of changes in IP-addresses (added, removed).

Discovery with external k/v stores was removed, and the `hostdiscovery`
package was removed in 00f9b23c3ae1614b648e9da1c6423c381f58153e, and
current code always pass a single IP-address (if any).

The Controller.pushNodeDiscovery method contained a check whether the
passed list of node-addresses was nil, however, this was never the
case because `Controller.handleNodeTableEvent` would unconditionally
[initialize a slice][2] (even if the event had a `nil` IP-address).

In case of a `nil` IP-address, this address would be;

- compared with the agent's advertiseAddr to determine if this is `self`
- encoded as a string, and sent as `NodeDiscoveryData.Address`. This string
  would be a literal `"<nil>"`.

While I don't know if there's other implementations that treat this information
different, looking at the [overlay implementation of `driver.DiscoverNew`][3],
it would check the field to be an empty string, and otherwise produce an
error (which wouldn't apply if it contained `"<nil>"`), and set the
driver's advertiseAddress to this value, effectively un-setting it
(`net.ParseIP()` silently ignores invalid values), or producing an error
(`netip.ParseAddr()` will error).

For completeness: `driver.DiscoverDelete` has no implementation in the
overlay driver, but both `driver.DiscoverNew` and `driver.DiscoverDelete`
are both implemented by the `remote` driver, which sends them to the
remote "as-is", so it will depend on the receiving end to handle.

So based on the overlay implementation, my assumption is that an empty
node-address should be considered _invalid_ (both for `DiscoverDelete`
and `DiscoverNew`), so let's return early if we encounter one, but log
a warning to discover cases where this happened.

[1]: https://github.com/moby/libnetwork/commit/ac4d7b61367e34ac341c2171d70d8b6a0a0fbecf
[2]: https://github.com/moby/moby/blob/5b53ddfcdd1c0721f875d3e237078ab7de891a57/libnetwork/agent.go#L905
[3]: https://github.com/moby/moby/blob/5b53ddfcdd1c0721f875d3e237078ab7de891a57/libnetwork/drivers/overlay/overlay.go#L92-L97

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on libnetwork/agent.go:
```diff
@@ -853,66 +853,56 @@ func (n *Network) handleDriverTableEvent(ev events.Event) {
-		value = event.Value
-		etype = driverapi.Delete
+		// TODO(thaJeztah): is it intentional for an "networkdb.UpdateEvent" event to send a "driverapi.Delete" notification?
+		d.EventNotify(driverapi.Delete, n.ID(), event.Table, event.Key, event.Value)
```
No idea if this was intentional, or a bug, so I left a TODO at least

thaJeztah commented:
@corhere @dperny ptal 🤗

corhere commented on libnetwork/agent.go:
```diff
@@ -853,66 +853,56 @@ func (n *Network) handleDriverTableEvent(ev events.Event) {
+		// TODO(thaJeztah): is it intentional for an "networkdb.UpdateEvent" event to send a "driverapi.Delete" notification?
+		d.EventNotify(driverapi.Delete, n.ID(), event.Table, event.Key, event.Value)
+	default:
+		// TODO(thaJeztah): are there other event-types? Should this log a warning?
```
[`events.Event` is an alias for `any`](https://pkg.go.dev/github.com/docker/go-events#Event) and [the event filter is very lenient](https://github.com/moby/moby/blob/2870559b2c23ed4803f10c566d955e33aa22a61c/libnetwork/networkdb/watch.go#L41-L86) so this function could theoretically be called with other event types. On the other hand, silently discarding is not an unreasonable approach for handling events that an event sink doesn't care about. Perhaps split the difference and log at debug level?

thaJeztah replied:
> Perhaps split the difference and log at debug level?

Works for me! I added a debug log

corhere reviewed (commented):
The handling of networkdb events would be so much simpler and less error-prone if there was a single exported `networkdb.Event` struct type with a `Type driverapi.EventType` field. All the conversion back-and-forth code would disappear.

thaJeztah commented:
> if there was a single exported networkdb.Event struct type with a Type driverapi.EventType field. All the conversion back-and-forth code would disappear.

Yes, perhaps that's something we could look into; was your intent to look at that as part of this PR, or for a follow-up?

corhere commented:
> was your intent to look at that as part of this PR, or for a follow-up?

I think it's in scope for this PR, but it's your PR so your call.

thaJeztah commented on libnetwork/agent.go:
```diff
@@ -831,66 +831,56 @@ func (n *Network) handleDriverTableEvent(ev events.Event) {
+		// TODO(thaJeztah): is it intentional for an "networkdb.UpdateEvent" event to send a "driverapi.Delete" notification?
+		d.EventNotify(driverapi.Delete, n.ID(), event.Table, event.Key, event.Value)
+	default:
+		log.G(context.TODO()).Debug("Network.handleDriverTableEvent: unhandled table event %v (%T) for driver %s", ev, ev, n.networkType)
```
Oh! Good catch from the linter; this should've been `Debugf`;

```
libnetwork/agent.go:843:3: printf: (*github.com/docker/docker/vendor/github.com/sirupsen/logrus.Entry).Debug call has possible Printf formatting directive %v (govet)
		log.G(context.TODO()).Debug("Network.handleDriverTableEvent: unhandled table event %v (%T) for driver %s", ev, ev, n.networkType)
		^
```

thaJeztah commented:
> I think it's in scope for this PR, but it's your PR so your call.

Ah, yes, I wanted to go look for those other changes, but didn't get further than rebasing the PR, then thought I'd push the rebase; will try to find some time and have a look again 😂

thaJeztah committed:
libnetwork/networkdb: remove opType and use driverapi.EventType

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/networkdb: eventDelegate.broadcastNodeEvent: add error-return

Make error-handling slightly more idiomatic, and log errors on the caller
side.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/networkdb: unify Event-types: export Event and add Type field

All events generated share the same underlying (event) type, but had to
be type-matched to detect the kind of event (create, update, delete).

Instead of defining different types for each event, we can share the
same type, and instead add a Type field, which holds the kind of event.

This patch:

- Adds a Type field to the event-type and exports it. The Type field
  uses the driverapi.EventType.
- Replaces uses of CreateEvent, UpdateEvent and DeleteEvent to use the
  Event type instead, and setting the Event.Type field accordingly.

Type-matching is still needed, due to the "go-events" package defining
the `Event` type as a generic `interface{}`, but logging was added for
cases where we get an unexpected type.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/networkdb: inline makeEvent

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on libnetwork/cmd/networkdb-test/dummyclient/dummyClient.go:
```diff
@@ -93,27 +87,19 @@ func handleTableEvents(tableName string, ch *events.Channel) {
+					delete(clientWatchTable[tableName].entries, event.Key)
+				default:
+					log.G(context.TODO()).Fatalf("Unexpected table event = %#v", event)
+					// FIXME(thaJeztah): old code would fall-through and call "delete" without key, or with the previously set key.
```
Added some "FIXME" comments in places where "technically" the behavior changes, but I think those cases where mostly theoretical (and they didn't seem to be _correct_ as well)

thaJeztah commented on libnetwork/networkdb/watch.go:
```diff
@@ -41,25 +48,26 @@ func (nDB *NetworkDB) Watch(tname, nid string) (*events.Channel, func()) {
-			if tname != "" && evt.Table != tname {
+			evt, ok := ev.(Event)
+			if !ok {
+				// FIXME(thaJeztah): verify this: the old code would fall-through, and return "true" ("Match")
```
Same here

thaJeztah commented on libnetwork/networkdb/watch.go:
```diff
@@ -41,25 +48,26 @@ func (nDB *NetworkDB) Watch(tname, nid string) (*events.Channel, func()) {
+				}
+				return true
+			default:
+				// FIXME(thaJeztah): verify this: the old code would fall-through, and return "true" ("Match")
 				return false
```
And here; although this code (before) was confusing, as it seemed to be returning random true/false for "other" type of events (i.e., if we didn't match it to any known type)

corhere replied:
Could be worse.
```go
matcher = events.MatcherFunc(func(ev events.Event) bool {
    evt, _ := ev.(Event)
    switch evt.Type {
    case driverapi.Create, driverapi.Delete, driverapi.Update:
        return true
    }
    return tname == "" || nid == ""
})
```

thaJeztah commented on libnetwork/networkdb/networkdb_test.go:
```diff
@@ -155,21 +156,18 @@ func testWatch(t *testing.T, ch chan events.Event, ev interface{}, tname, nid, k
+				assert.Check(t, is.Equal(nid, event.NetworkID))
+				assert.Check(t, is.Equal(key, event.Key))
+				assert.Check(t, is.Equal(value, string(event.Value)))
+			case driverapi.Delete:
+				assert.Check(t, is.Equal(tname, event.Table))
+				assert.Check(t, is.Equal(nid, event.NetworkID))
+				assert.Check(t, is.Equal(key, event.Key))
+			}
```
Separate case for `Delete` was still needed here, because the `Delete` would keep the `Value` of the event from what it was before (which would be `test_updated_value` in this case); I didn't want that tone part of the test (i.e., on `Delete` the content of `Value` doesn't matter, so isn't / shouldn't be checked).

thaJeztah reviewed (commented):
I pushed some commits that change  the Event type to have a `Type` field (which is a `driverapi.EventType`)

corhere reviewed (changes requested):
There are a couple problems with the networkdb changes that need to be fixed
